### PR TITLE
Add explicit locking to layer->user_tables relation

### PR DESCRIPTION
**Pull request by javitonino**
Fix for #11443 

There is a dedlock describe in this comment: https://github.com/CartoDB/cartodb/issues/11443#issuecomment-280626409. The important part:
1. Update controller updates the layer
2. Delete controller removes layers_user_tables
3. Update controller tries to add/remove layers_user_tables. Starts waiting for a lock
4. Delete controller tries to remove the layer. Starts waiting for a lock. Deadlock is inmediately detected and this request is killed (500).
5. Update completes successfully.

The problem is that the tables (`layers` and `layers_user_tables`) are locked in different orders in two operations. This change forces the `layers_user_table` to always be locked first, so deadlocks cannot occur, at the price of delaying the second operation until the first one finishes.

### CODE REVIEW

File: app/models/carto/layer.rb [hunk 1]
```diff
@@ -293,6 +294,13 @@ def depends_on?(user_table)
 
     private
 
+    # The table dependencies will only be updated after the layer. However, when deleting them, they need to be deleted
+    # before the model. This can cause deadlocks with simultaneous request to update and delete the model.
+    # This request a explicit lock to PostgreSQL so the tables are always accessed in the same order. #11443
+    def lock_user_tables
+      user_tables.lock.all
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Adding a new label this triggers this:

```sql
SELECT "user_tables".* FROM "user_tables" INNER JOIN "layers_user_tables" ON "user_tables"."id" = "layers_user_tables"."user_table_id" WHERE "layers_user_tables"."layer_id" IS NULL FOR UPDATE
```

Maybe it's too much but, do you think that checking `layer.id.present?` would be convenient?

Comment (javitonino) [inline app/models/carto/layer.rb]: Nice one!

Given that `layers_user_table.layer_id` is indeed nullable (...), I think it's worth the addition, to save the query which won't do anything.

**Comment by javitonino** [comment]
I'm optimist with this solving the deadlock, at least the one I located. After deploying this, I'll mark it as resolved and if it pops up again we can change the DB locks to see if it's the same one, or there are more places where something similar can happen. Unless, now that we know how this happens, it would be easier to fix.

**Comment by juanignaciosl** [comment]
👍

**Comment by javitonino** [comment]
Hard to do acceptance in this one. I would suggest to do a couple of tests to ensure the table dependencies are registered correctly (no regressions) and just test this live.

- [ ] Create a map with a dataset. Inmediately try to delete the dataset, you should be warned
- [ ] Create a map and edit the layer SQL to point it to another dataset. You should get a warning about the map if you try to delete the new dataset.
- [ ] Add and remove a layer. You should not get a warning when deleting the dataset of the deleted layer.

**Pull request merged**